### PR TITLE
[Offload] Initial support for registering offloading entries on COFF targets

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
This patch provides the initial support to allow handling the new
driver's offloading entries. Normally, the ELF target can emit varibles
at C-identifier named sections and the linker will provide a pointer to
the section. For COFF target, instead the linker merges sections
containing a `$` in alphabetical order. We thus can emit these variables
at sections and then emit two variables that are guaranteed to be sorted
before and after the others to traverse it. Previous patches
consolidated the handling of offloading entries so that this patch more
easily can handle mapping them to the appropriate section.

Ideally, the only remaining step to allow the new driver to run on
Windows targets is to accurately map the following `ld.lld` arguments to
their `llvm-link` equivalents. These are used inside the linker-wrapper,
so we should simply need to remap the arguments to the same
functionality if possible.
```
-o, -output
-l, --library
-L, --library-path
-v, --version
-rpath
-whole-archive, -no-whole-archive
```

I have not tested this at runtime as I do not have access to a windows
machine.

This patch was adapted from some initial efforts in
https://reviews.llvm.org/D137470.

### CODE REVIEW

File: llvm/lib/Frontend/Offloading/Utility.cpp [hunk 1]
```diff
@@ -62,35 +63,51 @@ void offloading::emitOffloadingEntry(Module &M, Constant *Addr, StringRef Name,
...
-                         /*isConstant=*/true, GlobalValue::ExternalLinkage,
-                         /*Initializer=*/nullptr, "__start_" + SectionName);
+  llvm::Triple Triple(M.getTargetTriple());
+
+  auto *ZeroInitilaizer =
+      ConstantAggregateZero::get(ArrayType::get(getEntryTy(M), 0u));
+  auto *EntryInit = Triple.isOSBinFormatCOFF() ? ZeroInitilaizer : nullptr;
+  auto *EntryType = Triple.isOSBinFormatCOFF()
+                        ? ZeroInitilaizer->getType()
+                        : ArrayType::get(getEntryTy(M), 0);
```
Comment (jdoerfert) [inline llvm/lib/Frontend/Offloading/Utility.cpp:82]: I don't see why we need the ternary here, aren't both options the same?

File: llvm/lib/Frontend/Offloading/Utility.cpp [hunk 2]
```diff
@@ -62,35 +63,51 @@ void offloading::emitOffloadingEntry(Module &M, Constant *Addr, StringRef Name,
...
   Entry->setAlignment(Align(1));
 }
 
 std::pair<GlobalVariable *, GlobalVariable *>
 offloading::getOffloadEntryArray(Module &M, StringRef SectionName) {
-  auto *EntriesB =
-      new GlobalVariable(M, ArrayType::get(getEntryTy(M), 0),
-                         /*isConstant=*/true, GlobalValue::ExternalLinkage,
-                         /*Initializer=*/nullptr, "__start_" + SectionName);
+  llvm::Triple Triple(M.getTargetTriple());
```
Comment (jdoerfert) [inline llvm/lib/Frontend/Offloading/Utility.cpp:75]: This should be in the llvm namespace.

**Review by jdoerfert — APPROVED** [review]
LG, two nits.

File: llvm/lib/Frontend/Offloading/Utility.cpp (hunk 1 above)
Comment (jhuber6) [reply llvm/lib/Frontend/Offloading/Utility.cpp:82]: One is a `EntryTy` the other is a `EntryTy[]`

Comment (jhuber6) [reply llvm/lib/Frontend/Offloading/Utility.cpp:81]: Actually you're right, forgot the initializer used the array type as well.

File: llvm/lib/Frontend/Offloading/Utility.cpp (hunk 2 above)
Comment (jhuber6) [reply llvm/lib/Frontend/Offloading/Utility.cpp:74]: I fixed the other occurrences, these should stay to separate the type as I prefer `llvm::Triple Triple` over `Triple TheTriple` or similar.

**Pull request merged**